cpp/tests/request.cpp

```diff
@@ -212,6 +212,49 @@ TEST_P(RequestTest, ProgressTagMulti)
     ASSERT_THAT(_recv[i], ContainerEq(_send[i]));
 }
 
+TEST_P(RequestTest, TagUserCallback)
+{
+  allocate();
+
+  std::vector<std::shared_ptr<ucxx::Request>> requests(2);
+  std::vector<ucs_status_t> requestStatus(2, UCS_INPROGRESS);
+
+  auto checkStatus = [&requests, &requestStatus](std::shared_ptr<void> data) {
+    auto idx = *std::static_pointer_cast<size_t>(data);
+    if (requests[idx] == nullptr) {
+      /**
+       * Unfortunately, we can't check the status this way if the request completes
+       * immediately, as `request[idx]` will only be assigned after completion, and thus,
+       * after the callback has been executed.
+       *
+       * TODO: find a better way to test this.
+       */
+      requestStatus[idx] = UCS_OK;
+    } else {
+      requestStatus[idx] = requests[idx]->getStatus();
+    }
+  };
+
+  auto sendIndex = std::make_shared<size_t>(0u);
+  auto recvIndex = std::make_shared<size_t>(1u);
+
+  // Submit and wait for transfers to complete
+  requests[0] = _ep->tagSend(_sendPtr[0], _messageSize, 0, false, checkStatus, sendIndex);
+  requests[1] = _ep->tagRecv(_recvPtr[0], _messageSize, 0, false, checkStatus, recvIndex);
+  waitRequests(_worker, requests, _progressWorker);
+
+  copyResults();
+
+  // Assert status was set before user callback is executed
+  for (const auto status : requestStatus)
+    ASSERT_THAT(status, UCS_OK);
+  for (const auto request : requests)
+    ASSERT_THAT(request->getStatus(), UCS_OK);
```

Comment:
So the logic here is that the callback runs with an updated request status (so we shouldn't observe `UCS_INPROGRESS` in `requestStatus`), and then we need to check that all the requests have a status of `UCS_OK`. Is that right?

If the first fails, it either means that the callback ran too late, or the messages somehow otherwise failed. (In which case the latter check also fails).

is that right?